File: backend/models/Workout.js
import mongoose from "mongoose";

const workoutSchema = new mongoose.Schema({
  name: {
    type: String,
    default: "Workout",
  },
  // 時間相關欄位
  startTime: {
    type: Date,
    required: true,
    default: Date.now, // 預設為創建時的時間
  },
  endTime: {
    type: Date,
    default: null, // 完成時設置
  },
  duration: {
    type: Number, // 單位：秒
    default: 0, // 完成時計算
  },
  // 訓練內容
  exercises: [
    {
      type: mongoose.Schema.Types.ObjectId,
      ref: "WorkoutSet",
    },
  ],
  // 其他欄位
  notes: {
    type: String,
    default: "",
  },
  photoUrl: {
    type: String,
    default: "",
  },
  status: {
    type: String,
    enum: ["in_progress", "completed"],
    default: "in_progress",
  },
});

// 自動計算訓練時長的方法
workoutSchema.methods.calculateDuration = function () {
  if (this.startTime && this.endTime) {
    // 計算時間差（毫秒）並轉換為秒
    this.duration = Math.floor((this.endTime - this.startTime) / 1000);
    return this.duration;
  }
  return 0;
};

// 完成訓練的方法
workoutSchema.methods.finish = function () {
  this.endTime = new Date();
  this.status = "completed";
  this.calculateDuration();
};

// 取消訓練的方法
workoutSchema.methods.cancel = function () {
  this.status = "cancelled";
};

export default mongoose.model("Workout", workoutSchema);

File: mobile/lib/models/workout.dart
import 'dart:convert';

class Workout {
  final String id;
  final String name;
  final String notes;
  final DateTime startTime;
  DateTime? endTime;
  int? duration;
  final String status; // 'in_progress', 'completed', 'cancelled'
  final List<String> exercises; // 存儲 WorkoutSet 的 ID

  Workout({
    required this.id,
    required this.name,
    required this.notes,
    required this.startTime,
    this.endTime,
    this.duration,
    required this.status,
    required this.exercises,
  });

  factory Workout.fromJson(Map<String, dynamic> json) {
    return Workout(
      id: json['_id'],
      name: json['name'] ?? '未命名訓練',
      notes: json['notes'] ?? '',
      startTime: DateTime.parse(json['startTime']),
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      duration: json['duration'],
      status: json['status'],
      exercises: List<String>.from(json['exercises'] ?? []),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'notes': notes,
      'startTime': startTime.toIso8601String(),
      'endTime': endTime?.toIso8601String(),
      'duration': duration,
      'status': status,
      'exercises': exercises,
    };
  }
}
